### ISSUE TITLE
Please add an overload with an IServiceProvider accessor to AddOData()

### ISSUE THREAD

**Issue by weitzhandler** [OData/AspNetCoreOData#9]
Hi,

First and foremost, thanks for the great work, the new way of setting up OData makes much more sense!

I'd like to request adding another overload of the `AddOData` extension set, such that its setupAction delegate comes with an `IServiceProvider` argument, so that we can resolve services (e.g. `IEdmModel` from IoC).

**Comment by weitzhandler** [comment]
Here's a potential implemenation:

```c#
public static IODataBuilder AddOData(this IServiceCollection services, Action<IServiceProvider, ODataOptions> setupAction)
{
    using var serviceProvider = services.BuildServiceProvider();
    using var serviceScope = serviceProvider.CreateScope();

    return services.AddOData(options => setupAction(serviceScope.ServiceProvider, options));
}
```

Alternatively, add another extension method to the `IOdataBuilder`, that allows setting the `IEdmModel` and maybe some other services, with an `IServiceProvider` accessor, for example:

```c#
services
.AddOData(options => { })
.UseEdmModel<MyEdmModel>();
```

Another options could be to always search for any services such `IEdmModel` or similar services that were pre-registered independently, when setting up OData.

**Comment by weitzhandler** [comment]
BTW, this was achievable in the old implementation, because `Startup.Configure` is dynamic, so we could register an `IEdmModel` then consume it in the following manner:

```c#
public void Configure(IApplicationBuilder app, IWebHostEnvironment env, IEdmModel edmModel)
{

app.UseEndpoints(routeBuilder =>
{
  routeBuilder.MapControllers();        


  app.UseEndpoints(routeBuilder =>
  {
    routeBuilder.MapControllers();        

    routeBuilder.MapODataRoute("API Route", urlOptions.ApiRoute, edmModel);
}
```

**Comment by weitzhandler** [comment]
@xuzhg I'm willing to contribute.

**Comment by weitzhandler** [comment]
This is important, because what if I need to inject services in the EDM model builder's constructor and want to have it all managed by the IoC?

**Comment by xuzhg** [comment]
@weitzhandler I'd love to see your contribution. I am working on the improvement recently, but focus on the routing template/convention parts. 

Your new extension method for `AddOData` seems reasonable to me. But, I can not understand `MapODataRoute` in https://github.com/OData/AspNetCoreOData/issues/9#issuecomment-708864509 . Such method is gone in this project. 
It means you don't need call MapODataRoute.

**Comment by weitzhandler** [comment]
> Your new extension method for `AddOData` seems reasonable to me. But, I can not understand `MapODataRoute` in [#9 (comment)](https://github.com/OData/AspNetCoreOData/issues/9#issuecomment-708864509) . Such method is gone in this project.
> It means you don't need call MapODataRoute.

Yeah, I saw, that's why I brought it in to demonstrate how it used to work. Anyway, OK, I'll make a PR.

**Comment by weitzhandler** [comment]
> Your new extension method for `AddOData` seems reasonable to me

I'll assume you meant `AddOData`, if you meant `UseEdmModel` please lemme know.

**Comment by weitzhandler** [comment]
@xuzhg 
Please have a look at my PR. I just rebased on master and resolved conflicts. Thank you.

**Comment by robertmclaws** [comment]
Hey folks, I changed the name of both this issue and the PR, just so they are less confusing. I saw the notification in my email and was concerned we were going back to the previous DI antipattern by putting IServiceProviders in constructors. But giving access to the IServiceCollection is not an antipattern. (Ideally, the services should be registered elsewhere, but since OData maintains its own container, this makes sense.

HTH!

**Comment by tomasfabian** [comment]
Hi,
I believe that this is the replacement for _MapODataRoute_. _IServiceCollection_  is already available as a paramater in _ConfigureServices_,

```C#
public class Startup
{
    public void ConfigureServices(IServiceCollection services)
    {
        services.AddOData(opt => opt.AddModel("odata", GetEdmModel()));
    }
}
```

**Comment by weitzhandler** [comment]
@robertmclaws @tomasfabian, indeed, it's the `IEdmModel` itself I want to resolve from the built service provider.

**Comment by rizi** [comment]
@weitzhandler thx, for the pullrequest, that's exactly what we need as well. @xuzhg any timeline for beta2?
br

**Comment by weitzhandler** [comment]
@rizi - custom IoC,
Some containers come with an adapter for MS.Ex.DI.

**Comment by tomasfabian** [comment]
I think that @robertmclaws is trying to avoid Service Locator antipattern. On the other hand we are forced to use so called "Poor mans DI" in case of IEdmModel and ODataBatchHandler (we have to manually instantiate them and their dependencies) outside of IoC container.
Now I see that options.AddModel can be called several times with different models. And it creates a new services provider for each of them. I think that in the result we have several composition roots.

@weitzhandler, @xuzhg what do you think about removing the IEdmModel parameter from _AddModel_ in this way?
```C#
services.AddOData(opt => opt.AddModel<EdmModel>("odata")); //singleton registration proposal

services.AddOData(opt => opt.AddModel("odata", builder => builder.AddService(Microsoft.OData.ServiceLifetime.Singleton, typeof(IEdmModel), typeof(EdmModel))));
```

In my opinion ODataBatchHandler should be in PerRequestScope not in Singleton. 

@rizi for example Autofac is able to remap the registered services:
```C#
      var builder = WebAssemblyHostBuilder.CreateDefault(args);
      builder.ConfigureContainer(new AutofacServiceProviderFactory(), containerBuilder =>
      {
        containerBuilder.Register(c => serviceModel)
          .As<IEdmModel>()
          .SingleInstance();
      }
```

**Comment by weitzhandler** [comment]
Personally I don't think it's a bad idea having a generic `AddModel`, I just think it doesn't have to be registered as singleton. Other than that we can simply create an `AddModel(Type modelType, Lifetime lifetime)` method, with an optional generic counterpart in the `ODataOptions` class, that internally registers that and loads it with the options configuration.

**Comment by tomasfabian** [comment]
@weitzhandler sure it doesn't have to be singleton, the singleton was meant as default
```C#
    public static void AddModel<TEdmModel>(string prefix, ServiceLifetime edmModelLifetime = ServiceLifetime.Singleton)
      where TEdmModel : IEdmModel
```

**Comment by tomasfabian** [comment]
@rizi I found out that there is a SetBuilderFactory to the rescue, which can be called before AddModel. 
```C#
options.SetBuilderFactory(() => new AutofacServiceProviderFactory().CreateBuilder(services));
```

Autofac.ContainerBuilder is not IContainerBuilder, but it is hopefully solvable.

**Comment by tomasfabian** [comment]
@weitzhandler your pull request is to add an overload with an IServiceCollection accessor to AddOData(). Wouldn't it be better to add another AddModel overload with IEdmModel registration as I suggested above?
  
What is the plan?

**Comment by weitzhandler** [comment]
What if someone wants to provide their model externally without having to register it? Additionally there might be additional service to rely on.
An alternative option, is to implement nothing and just have those cases manually call:

```c#
services
       .AddOptions<ODataOptions>()
       .Configure<IEdmModel>((options, model) => options.AddModel(model));
```

**Comment by tomasfabian** [comment]
They could use the classical way, to use method injection, if they don't want to register their model (_opt.AddModel("odata", GetEdmModel()_). It will internally register it for them. Additional services would be handled as before, IEdmModel is special, because it is required in the AddModel method. 

I'm fine with your last code snippet, because now I'm aware of this possibility. But isn't this a little bit tricky to figure out for everyone independently? If this approach will not be part of the official documentation. Injection/registration with AddModel is more straightforward in my opinion for fluent / irresistible APIs. 

Big advantage of your last code snippet is that it does not require nothing to implement, I totally agree with that. No unintentional bugs would be introduced, less code to maintain etc.

Thanks for the constructive communication up to this time!

**Comment by weitzhandler** [comment]
You're right, that's why I took the minimalistic approach of just adding another overload that gives you general access to the service provider.

So that you can:
```c#
service
  .AddOData((odataOptions, serviceProvider) => 
  {
     var model = serviceProvider.GetRequireService<IEdmModel>();
     odataOptions.AddModel(model);
  });
```

**Comment by tomasfabian** [comment]
This last code snippet looks like interception of the object creation for me. In my opinion it requires an additional step. The registration of the edmModel itself. 
I tried to not register it at all and I received an empty container, not a null value. This could probably lead to subtle bugs.
Are there really other types of "services" except of EdmModel to resolve in this case? What kind of services are we taking to account in this case?

I prefer the KISS approach in this case: adding only a new straightforward AddModel<TEdmModel> (or RegisterModel<TEdmModel>) to register the missing model and its dependencies.
